### DISCUSSION TITLE
Unexpected Unknown type after type narrowing a union type

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#4623] [Q&A]
A follow up to https://github.com/microsoft/pyright/issues/4621

I am wondering about a modification to the question using inherited `Generic` classes where the derived class is composed of additional `TypeVar` than the base class.

```python
from typing import Generic, TypeVar


class MyOtherBase:
    my_value: int


class MyOtherDerived(MyOtherBase):
    ...


########################################################


M = TypeVar("M", bound=MyOtherBase)
N = TypeVar("N", bound=MyOtherBase)


class MyBase(Generic[M]):
    my_attr: M

    def __init__(self, my_attr: M) -> None:
        super().__init__()
        self.my_attr = my_attr


class MyDerived(MyBase[M], Generic[M, N]):
    my_attr: M
    my_other_attr: N

    def __init__(self, my_attr: M, my_other_attr: N) -> None:
        super().__init__(my_attr=my_attr)
        self.my_other_attr = my_other_attr


########################################################


# after type narrowing N is inferred to be potentially Unknown
def my_function(my_classes_instance: MyBase[M] | MyDerived[M, N]):
    reveal_type(my_classes_instance)  # Type of "my_classes_instance" is "MyBase[M@my_function] | MyDerived[M@my_function, N@my_function]"
    if isinstance(my_classes_instance, MyDerived):
        reveal_type(my_classes_instance)  # Type of "my_classes_instance" is "MyDerived[M@my_function, Unknown] | MyDerived[M@my_function, N@my_function]"
        reveal_type(my_classes_instance.my_attr)  # Type of "my_classes_instance.my_attr" is "M@my_function"
        reveal_type(my_classes_instance.my_other_attr)  # Type of "my_classes_instance.my_other_attr" is "Unknown | N@my_function"


# functions as expected
def my_function_2(my_classes_instance: MyDerived[M, N]):
    reveal_type(my_classes_instance)  # Type of "my_classes_instance" is "MyDerived[M@my_function_2, N@my_function_2]"
    if isinstance(my_classes_instance, MyDerived):
        reveal_type(my_classes_instance)  # Type of "my_classes_instance" is "MyDerived[M@my_function_2, N@my_function_2]"
        reveal_type(my_classes_instance.my_attr)  # Type of "my_classes_instance.my_attr" is "M@my_function_2"
        reveal_type(my_classes_instance.my_other_attr)  # Type of "my_classes_instance.my_other_attr" is "N@my_function_2"


my_derived = MyDerived(MyOtherDerived(), MyOtherDerived())

my_function(my_derived)

```

It looks like the unshared `TypeVar` gets inferred as `Unknown` after type narrowing still? If I don't use a union type and just type it as `MyDerived[M, N]`, it behaves as expected.

I've taken the advice from https://github.com/microsoft/pyright/issues/4560. Have I formulated `MyDerived` and my hints properly? Anything I'm missing?

**Comment by erictraut** [comment]
Consider this simplified code, which removes the union from your code above.

```python
def func(x: MyBase[M]):
    reveal_type(x)
    if isinstance(x, MyDerived):
        reveal_type(x)
```

In this case, `x` starts with the type `MyBase[M]`. The narrowing logic for `isinstance` narrows it to `MyDerived`. Pyright can determine that the first type argument for `MyDerived` must be `M`, but there's no way for it to determine the second type argument (the one that corresponds to `N`) from the information it has been provided. It therefore fills in `Unknown`. So the narrowed type is `MyDerived[M@func, Unknown]`.

Your code is doing something unusual here that may indicate a misunderstanding of the way static typing works. It includes a union of `MyBase[M]` and `MyDerived[M, N]`. Why are you doing that? `MyDerived[M, N]` is a subtype of `MyBase[M]`, so you could simply use `MyBase[M]` to represent the same set of types. By adding `MyDerived[M, N]` to the union, you risk having `N` remain unsolved when calling a generic function that uses this union. Consider abandoning the union and using only `MyBase[M]`.

**Reply by ringohoffman** [reply]
What I am doing is something more like this:

```python
class MyContainer(Generic[M, N]):
    my_classes_instance: MyBase[M] | MyDerived[M, N]

    def calculate_some_value(self) -> int:
        if isinstance(self.my_classes_instance, MyDerived):
            return self.my_other_attr_value()
        return self.my_attr_value()

    def my_attr_value(self) -> int:
        ...
        return self.my_classes_instance.my_attr.my_value

    def my_other_attr_value(self) -> int:
        assert isinstance(self.my_classes_instance, MyDerived)
        ...
        return self.my_classes_instance.my_other_attr.my_value

```

I have an attribute that can be `MyBase` or `MyDerived`, and a special method that operates only on `MyDerived`. The real reason I want to differentiate `MyDerived` and `MyBase` is because I need to use attributes/values of `N` to do the calculation in `calculate_some_value` if `my_classes_instance` is a `MyDerived`.

While I guess I could add a parameter to this method that is `MyDerived` like this:

```python
    def my_other_attr_value(self, my_derived: MyDerived[M, N]) -> int:
        ...
        return my_derived.my_other_attr.my_value

```

It seems strange to to do that since `my_classes_instance` is already contained in `self`. I guess what I am really trying to say with this union is `MyDerived[M, N] | (MyBase[M] & NOT MyDerived[M, Any])`, as a sort of successive broadening of types. If `my_classes_instance` is not the more specific `MyDerived[M, N]`, then why do we need to solve for `N`?

Though I know it's not a one-to-one example, I feel like it's in the spirit of `overload` + `Literal[True]` & `Literal[False]` + a `bool` default: https://mypy.readthedocs.io/en/stable/literal_types.html#literal-types

**Reply by erictraut** [reply]
That's a pretty cumbersome approach. Your `MyContainer` class has two type arguments, but one of them applies only if the container happens to storing `MyDerived` instances. What value does `N` take on if your container is storing `MyBase` instances? 

If I were you, I'd consider different design patterns. One option is to create a class hierarchy for `MyContainer` that mirrors the hierarchy of the classes it is managing.

```python
class MyContainer(Generic[M]):
    my_classes_instance: MyBase[M]

    def calculate_some_value(self) -> int:
        return self.my_classes_instance.my_attr.my_value


class DerivedContainer(MyContainer[M], Generic[M, N]):
    my_classes_instance: MyDerived[M, N]

    def calculate_some_value(self) -> int:
        return self.my_classes_instance.my_other_attr.my_value
```

In any case, pyright is working as designed here. The `Unknown` type is expected in your original code snippet for the reasons I described above.
